Add About section

tulikavijay opened the pull request:
solves : [#202](https://github.com/WorldBrain/Memex/issues/202)

this is how it looks...
![webmemex](https://user-images.githubusercontent.com/25117249/34195751-2628431a-e585-11e7-93d8-3b3d713f8522.png)

I have some queries :
1. there's an acknowledgement section in the settings dashboard already, should i include it in the about page or not?(according to the original issue)
2. What should i use as an icon for the about page.
3. the changelog - i could acquire this much information. Should i add details and can you give me a link or so for the details. 

if there's changes to be made, please guide me. thanks :)

oliversauter commented:
Hey @tulikavijay 
Thanks for taking this on so quickly :)

Regarding the subsections "Vision, Changelog and acknowledgement" it was rather meant to put them as subpoints in the navigation bar and have "About" be the parent element.
Similar to like the early version of the import menu looked like: 
<img width="300" alt="screen shot 2017-12-20 at 12 51 05" src="https://user-images.githubusercontent.com/7870039/34205896-879f634a-e584-11e7-9543-7db6f0979638.png">
There we would have the options: 
- About
   - Vision
   - Tutorial
   - Acknowledgements
 
The Changelog itself would be a separate page. I can design this one, I just need to have the page and everything sorted in the sidebar.

tulikavijay commented:
Thanks for the reply. I misunderstood it a bit earlier but I got it now! I'll make a PR soon.

tulikavijay commented:
hey! if you don't mind, can you help me a bit with this? i added this section in NavLink.jsx : 
```    
{
route.name === 'About' &&
                children && (
                    <div className={styles.importSubItems}>
                        <div>1. Vision</div>
                        <div>
                            2. Changelog
                        </div>
                        <div>
                            3. Acknowledgements
                        </div>
                    </div>
                )}
```
I want to know what should i add in className?
Additionally I added Vision and Changelog folders...and added in routes - vision and changelog with children:About . Also some changes in the Head.js.
The frontend looks fine with the drop down and all , i don't know how to link all the components in the nav and the changes i have to make in About section if i have to.

Again thanks for the help, I am relatively new to react specially large projects.

poltak commented:
@tulikavijay the `className` prop on React elements corresponds to the `class` attribute in HTML elements. So specifying `<div className="foo">` in your React render method should result in `<div class="foo">` being produced for rendering in the browser. 

The `styles.importSubItems` is just applying the class `importSubItems` to that `<div>`, as defined in the CSS file that `styles` is imported from - we're using CSS modules to import the CSS classes like that in the React code. So you could go into that CSS file and change the rules in the `.importSubItems` selector's block to change, as you do normally in CSS styling.

Assuming you have the routes set up for each of the sub menu items (vision, changelog, ack), you can `import { Link } from 'react-router'` and use that component instead of `<div>` to make those components link to the specific routes (check the rest of the code, or the docs, for examples of using `Link` component).

tulikavijay commented:
Thanks for the reply and thanks for being patient!
The navbar looks good, I'll add in a screenshot. I have linked the pages too - vision, acknowledgement and changelog.
I have few more questions :
1. What should I use as an icon for 'About'.
2. I previously added the about folder which included index.jsx and css, now when I click the about link the drop down appears as well as the about page(with subsections added intially) - I can't remove it in the route.js and head.js, what should I do?
3. When the links open - the drop down doesn't stay for the newly loaded page.

poltak commented:
> 2. I previously added the about folder which included index.jsx and css, now when I click the about link the drop down appears as well as the about page(with subsections added intially) - I can't remove it in the route.js and head.js, what should I do?
> 3. When the links open - the drop down doesn't stay for the newly loaded page.


Not sure exactly what you mean here. Can you commit some code with how you've added the sub-links in nav (is that the "drop down"?)? 

Looking at the routes file, there is only a single-level so not sure how you would add the second-level links, without changing the data structure slightly (maybe an opt. `subroutes` array on route definitions would be nice, adding support in the [code that uses the `routes` data](https://github.com/WorldBrain/WebMemex/blob/master/src/options/components/navigation/index.jsx#L29)). Looking at the current import nav, it does it by doing a [check within the `NavLink` view](https://github.com/WorldBrain/Memex/blob/master/src/options/components/navigation/NavLink.jsx#L47), although I __would not__ recommend doing it like this as it is a bit of a hack and not a concern of `NavLink` view (IIRC this code was never reviewed before merge). 

Also, is each sub nav link meant to be clickable, or just for display? And are they meant to be separate pages, or just all appear in the about page?

tulikavijay commented:
Sorry for being confusing, the drop-down is the sub-links in the nav. the sub nav links are meant to be clickable and i think the pages are meant to be seperate.

Also i did make changes in the Navlink view (which you told me was a hack) and my questions earlier were regarding that.(And the PR i just submitted refers to the previous questions and it contains changes in Navlink view)
I thought about the alternative( i.e. opting subroutes array on route definitions, adding support in the code that uses the routes data ), tried the alternative and the code became a bit of a mess...it ran into an error so i couldn't commit it and push it.(thus the PR)

this is what i tried : 
 ### route.js : 
```
 {
            name: 'About',
            pathname: '/about',
            component: About,
            icon: 'info',
            subItems:[
            {
                name: 'Vision',
                pathname: '/vision',
                component: Vision,
                children: About,
                hideFromSidebar: true,
                subLink: true,
            },
            {
              ...
            },
            {
              ...
            }
            ]
         },
```
### index.jsx

```
return this.props.routes
            .filter(route => !route.hideFromSidebar)
            .map((route, idx) => ({
               if(!route.subSections){
                 <NavLink route={route} key={idx} state={state}>
                     {this.isActive(route)}
                 </NavLink>
               }
              else{
                  return route.subSections.map((child,idx)=>(
                      <NavLink route={child} key={idx} state={state}>
                          {this.isActive(child)}
                      </NavLink>
                   ))
              }
           }
          ))

```
### Navlink.js
```
{route.subLink &&(  // subLink is present in Head.js for subItems
  <div className="">
  <Link to={route.pathname}>{route.name}</Link>
  </div>
)}
```

poltak commented:
@tulikavijay With those code snippets, in the `index.jsx` you're doing stuff with `route.subSections` but no property exists with this name in the static data in your `routes.js` snippet. Rather there is `subItems` array in there. This could have just been a typo though?

Regardless, having nested `Array.prototype.map` calls is going to lead to nested arrays being returned from `renderNavLinks` method. Rather I would be passing in sub nav links as `props.children` on `NavLink` - currently `props.children` is used for active state which is really weird (`props.children` is generally reserved for string/JSX-element/s in react, not bools - active state should go on a custom prop, maybe like `props.isActive` - not your fault though, this is how this code already is). So maybe something like:

### index.jsx

```js
    renderNavSubLinks = (routes = []) => 
        routes.map((route, i) => <SubNavLink key={i} {...route} />)

    renderNavLinks() {
        ... 
        return this.props.routes
            .filter(route => !route.hideFromSidebar)
            .map((route, idx) => (
                <NavLink route={route} key={idx} state={state} isActive={this.isActive(route)}>
                    {this.renderNavSubLinks(route.children)}
                </NavLink>
            ))
    }
```

`<SubNavLink>` being a super simple view to represent the sublinks, and `route.children` is akin to your optional `subItems` array in your route.js snippet - I would probably go with the name "children", to be really abstract, or even just "subroutes", to distance it from concepts like "links" or "sections" that are more UI-specific (`routes.js` definitions also used to set up the router as well as nav UI structure).

Your code looks pretty good so far though :)

Going by the OP in #202 - the accordion idea - they're all just a part of the about page rather than separate pages, and they also shouldn't be clickable (@oliversauter correct me if I'm wrong), so `<Link>` is wouldn't be appropriate here (my bad suggestion) - maybe `SubNavLink` can simply be like the current sub nav items for the imports route (if you think that structure is good - I haven't really looked at much of this code)

oliversauter commented:
> they're all just a part of the about page rather than separate pages, and they also shouldn't be clickable

Ah there must be a misunderstanding then. I meant to have them all as separate pages with links from each subpoint. Structure:
- About
  - Vision
  - Changelog
  - Acknowledgements

tulikavijay commented:
For some reason the sub nav is not rendering...
Also I added `exact` prop for `<Route>` in option.js and head.js to not render the about page( a bit hacky and not working)
